src/sig_variant.zig

```diff
@@ -643,60 +649,152 @@ pub fn createSigVariant(
+            var wg = std.Thread.WaitGroup{};
+
+            const cpu_count = @max(1, std.Thread.getCpuCount() catch 1);
+            const n_workers = @min(cpu_count, n_elems);
+            const Signature = @This();
+
+            for (0..n_workers) |_| {
+                spawnTaskWg(&wg, struct {
+                    fn run(_msgs: [][]const u8, _dst: []const u8, _pks: []const *PublicKey, _pks_validate: bool, _sigs: []const *Signature, _sigs_groupcheck: bool, _rands: [][]const u8, _rand_bits: usize, _pool: *MemoryPool, _atomic_counter: *AtomicCounter, _atomic_valid: *AtomicError) void {
+                        const pairing_buffer = _pool.getPairingBuffer() catch {
+                            // .release will publish the value to other threads
+                            _atomic_valid.store(BLST_FAILED_PAIRING, AtomicOrder.release);
+                            return;
+                        };
+                        defer {
+                            _pool.returnPairingBuffer(pairing_buffer) catch {
+                                // .release will publish the value to other threads
+                                _atomic_valid.store(BLST_FAILED_PAIRING, AtomicOrder.release);
+                            };
+                        }
+                        var pairing = Pairing.new(&pairing_buffer[0], pairing_buffer.len, hash_or_encode, &_dst[0], _dst.len) catch {
+                            // .release will publish the value to other threads
+                            _atomic_valid.store(BLST_FAILED_PAIRING, AtomicOrder.release);
+                            return;
+                        };
+
+                        // the most relaxed atomic ordering
+                        var local_count: usize = 0;
+                        while (_atomic_valid.load(.monotonic) == 0) {
+                            // this uses @atomicRmw internally and returns the previous value
+                            // acquired then release which publish value to other thread
+                            const counter = _atomic_counter.fetchAdd(1, AtomicOrder.acq_rel);
+                            if (counter >= _msgs.len) {
+                                break;
+                            }
+                            pairing.mulAndAggregate(&_pks[counter].point, _pks_validate, &_sigs[counter].point, _sigs_groupcheck, &_rands[counter][0], _rand_bits, &_msgs[counter][0], _msgs[counter].len, null) catch {
+                                // .release will publish the value to other threads
+                                _atomic_valid.store(c.BLST_VERIFY_FAIL, .release);
+                                return;
+                            };
+                            local_count += 1;
+                        }
+
+                        if (local_count > 0 and _atomic_valid.load(.monotonic) == 0) {
+                            pairing.commit();
+                            if (!pairing.finalVerify(null)) {
```

Comment:
one difference to Rust binding is it does not do `finalVerify()` in every thread, instead it send pairing back to the main/issuing thread and `merge()` pairings, then do exactly one single `finalverify()`

note that `Pairing.merge()` was not implemented in this PR so may give it a try later to see if performance is better